code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+<https://video.uva.nl/media/Platforms2021%20Typing%20Python/0_d7kh0wak>
+
+
+    def sum(items):
+        total = 0
+        for item in items:
+            total += item
+        return total
+
+
+Python's approach is simple, we'll just run the code and see if it works. If `items` can be summed, then great let's do that. This all works:
+
+
+    sum([1, 2])
+    sum([1, 2.0])
+    sum({1, 2, 3})
+
+
+But this does **not**:
+
+
+    sum(["hello", 1])
+
+
+And worse yet, we won't know that it does not work until this code is actually run. If the code is not properly tested, then running this function might not happen until its shipped to the client. In which case... **nightmares**.
+
+
+#### C
+
+Okay, but what about other languages? Remember C?
+
+![embed](https://api.eu.kaltura.com/p/120/sp/12000/embedIframeJs/uiconf_id/23449960/partner_id/120?iframeembed=true&playerId=kaltura_player&entry_id=0_4la7ngbg&flashvars[streamerType]=auto&amp;flashvars[localizationCode]=en_US&amp;flashvars[leadWithHTML5]=true&amp;flashvars[sideBarContainer.plugin]=true&amp;flashvars[sideBarContainer.position]=left&amp;flashvars[sideBarContainer.clickToClose]=true&amp;flashvars[chapters.plugin]=true&amp;flashvars[chapters.layout]=vertical&amp;flashvars[chapters.thumbnailRotator]=false&amp;flashvars[streamSelector.plugin]=true&amp;flashvars[EmbedPlayer.SpinnerTarget]=videoHolder&amp;flashvars[dualScreen.plugin]=true&amp;flashvars[hotspots.plugin]=1&amp;flashvars[Kaltura.addCrossoriginToIframe]=true&amp;&wid=0_sucghmat)
+
+<https://video.uva.nl/media/Platforms2021%20Typing%20C/0_4la7ngbg>
+
+
+    int sum(int items[], int n) {
+        int total = 0;
+        for (int i = 0; i < n; i++) {
+            total += items[i];
+        }
+        return total;
+    }
+
+
+C takes a different approach, put a concrete type in front of everything and check it when trying to compile. That way we'll know up front whether the code will even run. Because this:
+
+
+    float array[] = {3.0, 4.0, 5.0};
+    sum(array, 3);
+
+
+Will nicely throw a compile error. No chance that this code reaches the end user's desk. 
+
+But wait, floats can be summed right? Well, tough luck. You'll need to write a new function for floats.
+
+> For the curious, there are ways to escape C's type system through the use of casting and pointers. Most notably through the use of `void` pointers.
```

Comment:
"escape C’s type system" is misschien niet snel te begrijpen -> "avoid strict type checking in C, too"

Het voorbeeld van de void pointers (dat jij bedoelt) is denk ik niet makkelijk te googlen op basis van hoe je het hier beschrijft, dus voor de nieuwsgierigen misschien een dead end.

Reply:
Dit is de eerste google hit voor void pointer C: https://www.geeksforgeeks.org/void-pointer-c-cpp/

Lijkt me een redelijke goede uitleg